cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/MetadataClient.java

```diff
@@ -1,126 +1,163 @@
-      // Cruise Control always fetch metadata for all the topics.
-      _metadata.needMetadataForAllTopics(true);
-      int version = _metadata.requestUpdate();
-      long remaining = timeout;
-      Cluster beforeUpdate = _metadata.fetch();
-      boolean isMetadataUpdated = _metadata.version() > version;
-      while (!isMetadataUpdated && remaining > 0) {
-        _metadata.requestUpdate();
-        long start = _time.milliseconds();
-        _networkClient.poll(remaining, start);
-        remaining -= (_time.milliseconds() - start);
-        isMetadataUpdated = _metadata.version() > version;
-      }
-      if (isMetadataUpdated) {
-        if (LOG.isDebugEnabled()) {
-          LOG.debug("Updated metadata {}", _metadata.fetch());
-        }
-        if (MonitorUtils.metadataChanged(beforeUpdate, _metadata.fetch())) {
+    if (_time.milliseconds() >= _lastSuccessfulUpdateMs + _metadataTTLMs) {
+      try {
+        // Cruise Control always fetches metadata for all the topics.
+        Cluster refreshedCluster = doRefreshMetadata(timeoutMs);
+        _lastSuccessfulUpdateMs = _time.milliseconds();
+        _version += 1;
+        LOG.debug("Updated metadata {}", _cluster);
+        if (MonitorUtils.metadataChanged(_cluster, refreshedCluster)) {
           _metadataGeneration.incrementAndGet();
+          _cluster = refreshedCluster;
         }
-      } else {
+      } catch (ExecutionException | TimeoutException | InterruptedException e) {
+        LOG.warn("Exception while updating metadata", e);
         LOG.warn("Failed to update metadata in {}ms. Using old metadata with version {} and last successful update {}.",
-                 timeout, _metadata.version(), _metadata.lastSuccessfulUpdate());
+            timeoutMs, _version, _lastSuccessfulUpdateMs);
       }
     }
-    return new ClusterAndGeneration(_metadata.fetch(), _metadataGeneration.get());
+    return new ClusterAndGeneration(_cluster, _metadataGeneration.get());
   }
 
-  /**
-   * Close the metadata client. Synchronized to avoid interrupting the network client during a poll.
-   */
-  public synchronized void close() {
-    _networkClient.close();
+  private Cluster doRefreshMetadata(int timeoutMs) throws InterruptedException, ExecutionException, TimeoutException {
+    // We use both AdminClient timeout options and `future.get(timeout)`
+    // The former ensures the timeouts are enforced in the brokers
+    int remainingMs = timeoutMs;
+    long startMs = _time.milliseconds();
+
+    Set<String> topicNames = _adminClient.listTopics(new ListTopicsOptions().timeoutMs(remainingMs).listInternal(true))
+        .names().get(remainingMs, TimeUnit.MILLISECONDS);
+    long endMs = _time.milliseconds();
+    long elapsedTimeMs = endMs - startMs;
+    remainingMs -= elapsedTimeMs;
+
+    DescribeClusterResult result = _adminClient.describeCluster(new DescribeClusterOptions().timeoutMs(remainingMs));
```

Comment:
Yes, our flow of `listTopics->describeCluster->describeTopics` means that each call could return different, stale information.

Thankfully, I think the calls are very tightly related to one another. I envision a small amount of possible scenarios:
* node in describeCluster but not in topic placements - likely no replicas have been placed in the node yet
* node NOT in describeCluster but in topic placements - this is obviously inconsistent and something is wrong. I proposed retrying here
* topic from listTopics is deleted and describeTopics returns an error for it - we should just catch this I think

Reply:
> node in describeCluster but not in topic placements - likely no replicas have been placed in the node yet

I suspect that this might also indicate one of the following scenarios:
* Metadata staleness of the broker that responds to `listTopics` -- e.g. 
   1. The cluster has 2 brokers: `B1` and `B2`.
   2. `B1` has replicas, but `B2` has no replicas.
   3. A topic is created, which puts a replica on `B2`.
   4. `listTopics` request goes to `B1`, which returns topics excluding the new topic, because it hasn't heard about topic creation, yet.
* There was a replica on this node, but it was moved away from it after `describeCluster`, but before `describeTopics`.

> node NOT in describeCluster but in topic placements - this is obviously inconsistent and something is wrong. I proposed retrying here

As you correctly pointed out, this could be due to multiple reasons, including ongoing interactions of broker addition, topic creation, replica moves, and transient broker failures. In case this inconsistency occurs during an ongoing replica move, since this process relies on retrieval of metadata to make progress, I suspect that a retry logic may have adverse affects (e.g. prolonged inconsistencies in retrieved nodes and topic placements due to lack of guarantees on which broker receives the request) on the progress of execution.

> topic from listTopics is deleted and describeTopics returns an error for it - we should just catch this I think

I agree that this is also a possible outcome, and catching this (and retrying) sounds reasonable to me. Again though, in clusters with high topic churn (deletion/creation), this may have similar adverse affect on execution as described on the prev. item.

I guess my other concern is, when we put pieces of metadata together, we are always under the risk of generating an inconsistent world view, not necessarily shared by any broker at a given time.